src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+    {
+        CHANNEL_OP* channel_op = CONTAINING_RECORD(entry, CHANNEL_OP, anchor);
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_096: [ set the result of the operation to CHANNEL_CALLBACK_RESULT_ABANDONED. ]*/
+        channel_op->result = CHANNEL_CALLBACK_RESULT_ABANDONED;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_097: [ call threadpool_schedule_work with execute_callbacks as work_function. ]*/
+        if (threadpool_schedule_work(channel_internal_ptr->threadpool, execute_callbacks, channel_op) != 0)
+        {
+            LogError("Failure in threadpool_schedule_work(execute_callbacks, channel_op)");
+        }
+    }
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_100: [ channel_internal_close shall call srw_lock_release_exclusive. ]*/
+    srw_lock_release_exclusive(channel_internal_ptr->lock);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CHANNEL_INTERNAL), channel_internal_create_and_open, THANDLE(THREADPOOL), threadpool)
+{
+    THANDLE(CHANNEL_INTERNAL) result = NULL;
+
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_098: [ channel_create shall call srw_lock_create. ]*/
+    SRW_LOCK_HANDLE lock = srw_lock_create(false, "channel");
+    if (lock == NULL)
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+        LogError("Failure in srw_lock_create(false, \"channel\")");
+    }
+    else
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_078: [ channel_create shall create a CHANNEL_INTERNAL object by calling THANDLE_MALLOC with channel_internal_dispose as dispose.]*/
+        THANDLE(CHANNEL_INTERNAL) channel_internal = THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose);
+        if (channel_internal == NULL)
+        {
+            /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+            LogError("Failure in THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose)");
+        }
+        else
+        {
+            CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_internal);
+
+            channel_internal_ptr->lock = lock;
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_080: [ channel_create shall store given threadpool in the created CHANNEL_INTERNAL. ]*/
+            THANDLE_INITIALIZE(THREADPOOL)(&channel_internal_ptr->threadpool, threadpool);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_084: [ channel_create shall call DList_InitializeListHead. ]*/
+            DList_InitializeListHead(&channel_internal_ptr->op_list);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_086: [ channel_create shall succeed and return the created THANDLE(CHANNEL). ]*/
+            THANDLE_INITIALIZE_MOVE(CHANNEL_INTERNAL)(&result, &channel_internal);
+            goto all_ok;
+        }
+        srw_lock_destroy(lock);
+    }
+all_ok:
+    return result;
+}
+
+static void execute_callbacks(void* context)
```

Comment:
>execute_callbacks [](http://example.com/codeflow?start=12&length=17)

I don't think execute_callbacks can access any of the fields of channel_op in the absence of locks.

This is (at least) because in cancel_op the field "result" is accessed under a lock.

Reply:
The lock is actually just needed while removing the operation from the list, does not need to be held while accessing result.

Reply:
Moving release so that only DList operation is under lock.